src/utils.ts

```diff
@@ -747,3 +755,247 @@ export const messageSetPagination = <StreamChatGenerics extends ExtendableGeneri
     return messagePaginationLinear(params);
   }
 };
+
+/**
+ * A utility object used to prevent duplicate invocation of channel.watch() to be triggered when
+ * 'notification.message_new' and 'notification.added_to_channel' events arrive at the same time.
+ */
+const WATCH_QUERY_IN_PROGRESS_FOR_CHANNEL: Record<string, Promise<QueryChannelAPIResponse> | undefined> = {};
```

Comment:
All of the utilities below this are snagged from [here](https://github.com/GetStream/stream-chat-react/blob/master/src/utils/getChannel.ts) and (here)[https://github.com/GetStream/stream-chat-react/blob/master/src/components/ChannelList/utils.ts].